base/cvd/cuttlefish/host/commands/cvd/server_command/create.cpp

```diff
@@ -0,0 +1,337 @@
+      const cvd_common::Args& subcmd_args, const cvd_common::Envs& envs,
+      const RequestWithStdio& request);
+
+  static void MarkLockfiles(std::vector<InstanceLockFile>& lock_files,
+                            const InUseState state);
+  static void MarkLockfilesInUse(std::vector<InstanceLockFile>& lock_files) {
+    MarkLockfiles(lock_files, InUseState::kInUse);
+  }
+
+  InstanceManager& instance_manager_;
+  CommandSequenceExecutor& command_executor_;
+};
+
+void CvdCreateCommandHandler::MarkLockfiles(
+    std::vector<InstanceLockFile>& lock_files, const InUseState state) {
+  for (auto& lock_file : lock_files) {
+    auto result = lock_file.Status(state);
+    if (!result.ok()) {
+      LOG(ERROR) << result.error().FormatForEnv();
+    }
+  }
+}
+
+Result<bool> CvdCreateCommandHandler::CanHandle(
+    const RequestWithStdio& request) const {
+  auto invocation = ParseInvocation(request.Message());
+  return Contains(CmdList(), invocation.command);
+}
+
+Result<selector::LocalInstanceGroup> CvdCreateCommandHandler::GetOrCreateGroup(
+    const std::vector<std::string>& subcmd_args, const cvd_common::Envs& envs,
+    const RequestWithStdio& request) {
+  using CreationAnalyzerParam =
+      selector::CreationAnalyzer::CreationAnalyzerParam;
+  const auto& selector_opts =
+      request.Message().command_request().selector_opts();
+  const auto selector_args = cvd_common::ConvertToArgs(selector_opts.args());
+  CreationAnalyzerParam analyzer_param{
+      .cmd_args = subcmd_args, .envs = envs, .selector_args = selector_args};
+
+  auto analyzer = CF_EXPECT(instance_manager_.CreationAnalyzer(analyzer_param));
+  auto group_creation_info = CF_EXPECT(analyzer.ExtractGroupInfo());
+
+  std::vector<InstanceLockFile> lock_files;
+  for (auto& instance : group_creation_info.instances) {
+    CF_EXPECT(instance.instance_file_lock_.has_value(),
+              "Expected instance lock");
+    lock_files.emplace_back(std::move(*instance.instance_file_lock_));
+  }
+
+  auto groups = CF_EXPECT(instance_manager_.FindGroups(selector::Query(
+      selector::kGroupNameField, group_creation_info.group_name)));
+  CF_EXPECTF(groups.size() <= 1,
+             "Expected no more than one group with given name: {}",
+             group_creation_info.group_name);
+  // When loading an environment spec file the group is already in the database
+  // in PREPARING state. Otherwise the group must be created.
+  if (groups.empty()) {
+    groups.push_back(
+        CF_EXPECT(instance_manager_.CreateInstanceGroup(group_creation_info)));
+  } else {
+    auto& group = groups[0];
+    CF_EXPECTF(group.Instances().size() == group_creation_info.instances.size(),
+               "Mismatch in number of instances from analisys: {} vs {}",
+               group.Instances().size(), group_creation_info.instances.size());
+    // The instances don't have an id yet
+    for (size_t i = 0; i < group.Instances().size(); ++i) {
+      auto& instance = group.Instances()[i];
```

Comment:
out of curiosity, do we have any high level overview comment around the relationship between `instances`, instances_info` and `instance_file_lock` concepts, they seem to map 1:1, but still I cannot get the full picture from the code only, wondering whether we could add some comment about it if it does not exist. thanks.

Reply:
Those need a refactor, not a comment.

- The cvd::Instance is the representation of an instance for cvd, it's the object stored in the database and what should be used almost everywhere.
- An instance_info was initial information about an instance before it was added to the database, the class doesn't exist anymore, but there is InstanceGroupInfo which is the same thing for groups. A LocalInstanceGroup is created from a InstanceGroupInfo.
- instance_file_lock is missnamed. It represents a lock on the instance id (not the instance itself) so that different instances don't use the same id.

I have done some refactors around this, but there is still some more to do.

Reply:
I see, appreciate the context, thanks!